### PR TITLE
Configuring and overriding html tags

### PR DESCRIPTION
**Pull request by dhartunian**

### CODE REVIEW

**Comment by dhartunian** [comment]
Def not "complete" by any means but would like some early feedback on design!

**Comment by emhoracek** [comment]
ooh, a RenderContext. Interesting!

**Comment by emhoracek** [comment]
I wonder if the Substitutions should be part of the RenderContext as well? Then the RenderContext is all the things that are needed to render the template?

**Comment by dhartunian** [comment]
@emhoracek good idea. I'll give it a shot

**Comment by dhartunian** [comment]
here's a stab at it...some function calls in the code get a bit messy looking because i just replaced things with `(substitutions rc)` in a bunch of places...

File: test/Spec.hs [hunk 1]
```diff
@@ -116,7 +123,9 @@ spec = hspec $ do
           mempty,
           M.fromList [(["default", "x"], parse "hello")
                      ,(["foo", "bar", "baz"], parse "<apply-content/>")]) `shouldRender` "hello"
-
+    it "should allow overriden Html tags" $ do
+      ("<html><div></div></html>",
+       RenderContext mempty (Overrides mempty (HS.fromList ["div"])) (fills [("div", text "notadivatall")])) `shouldRenderCustom` "<html>not a div at all</html>"
```
Comment (dbp) [inline test/Spec.hs]: How is this passing? Where do the extra spaces come from?!!

**Comment by dbp** [comment]
This looks pretty good. I think my two concerns are:

1. I'd like to see more tests! In particular, I'm curious about the nesting working out (ie, you showed that `<html><div/></html>` works out, but does `<html><custom><div/></custom></html>` as well?) - not that I doubt it, just that it's something that could absolutely go wrong, so good to have a test.

2. I'm not sure about `RenderContext` - in particular, it seems like the overrides aren't actually used at render-time (as, indeed, `renderWith` makes a new default set of overrides, which are empty), which I _think_ is right, but makes me think that maybe the overrides don't need to be in it?

And of course, we should use `customPlainNodes`!

File: test/Spec.hs (hunk 1 above)
Comment (emhoracek) [inline test/Spec.hs]: @dbp this is maybe a flaw in that should Render helper that strips out the spaces of   the rendered and the expected text

**Comment by emhoracek** [comment]
@dhartunian I'm going to try moving the overrides into an argument to `parse` and see what that looks like.

**Comment by emhoracek** [comment]
I removed the RenderContext and moved overrides into an argument to parse/mk, and it works! Not sure about whether I should merge that into this branch as it undoes a whole lot of @dhartunian's work (SORRY!! :grimacing:). Also I did like that adding a RenderContext collapsed a couple of the args to `process`, the type signature of which looks kinda ridic now. Maybe instead, there's a `processContext`?? (Thinking of an article I can't find right now that encouraged "local" datatypes -- could be handy here!)

**Pull request merged**